### PR TITLE
Add best pipelines

### PR DESCRIPTION
**Pull request by orionw**
Features
- Checks database for pipelines and grabs the best pipeline for every seed problem we have data on.
- Updates and places pipeline in location for submission
- Checks and prints out how many of our pipelines beat MIT/EXlines
- Updates to newest docker image for d3m

### CODE REVIEW

File: run_tests.sh [hunk 1]
```diff
@@ -3,8 +3,8 @@
 pip3 install .
 mv byudml tmp_byudml
 reset
-python3 primitive_jsons/generate_primitive_jsons.py
-python3 pipelines/generate_pipelines.py
+python3 submission/pipelines/generate_pipelines.py
+python3 submission/primitive_jsons/generate_primitive_jsons.py
```
Comment (bjschoenfeld) [inline run_tests.sh]: Let's switch the order back on these two lines. Generating the primitive json has validation code that should be run before we try to make a pipeline with our primitives.

File: setup.py [hunk 2]
```diff
@@ -1,7 +1,11 @@
 from setuptools import setup, find_packages
 
 from byudml import __version__
+from byudml import __imputer_path__
+from byudml import __metafeature_path__
 
+# make sure this is a valid entrypoint
+assert "d3m.primitives.metalearning.metafeature_extractor.BYU" == ("d3m.primitives." + ".".join(__metafeature_path__.split(".")[2:]))
```
Comment (bjschoenfeld) [inline setup.py]: Having this line means that the python path is no longer single sourced.

**Comment by bjschoenfeld** [comment]
Why do we need the `best_pipelines.csv` file? I thought this was going to look straight at the database.

File: submission/pipelines/generate_pipelines.py [hunk 3]
```diff
@@ -396,33 +404,171 @@ def remove_digests(
...
+        id_matches = check_step["primitive"]["id"] == step["primitive"]["id"]
+        if not id_matches:
+            step["primitive"]["id"] = check_step["primitive"]["id"]
+        version_matches = check_step["primitive"]["version"] == step["primitive"]["version"]
+        if not version_matches:
+            step["primitive"]["version"] = check_step["primitive"]["version"]
+        # make sure the digest exists
+        digests_match = "digest" in step["primitive"] and check_step["primitive"]["digest"] == step["primitive"]["digest"]
+        if not digests_match:
+            step["primitive"]["digest"] = check_step["primitive"]["digest"]
```
Comment (bjschoenfeld) [inline submission/pipelines/generate_pipelines.py]: We need to remove digests, not update them. Other teams may update their primitives after we submit our primitives and pipelines. If they do, the digests will change and our submitted pipelines become invalid.

File: setup.py (hunk 2 above)
Comment (orionw) [reply setup.py]: Oh, that was debugging code.  Let me take that out

File: submission/pipelines/generate_pipelines.py (hunk 3 above)
Comment (orionw) [reply submission/pipelines/generate_pipelines.py]: Yes, totally right.  I'll remove those lines

File: submission/pipelines/generate_pipelines.py [hunk 4]
```diff
@@ -1,6 +1,7 @@
 import json
 import os
-
+import pymongo
```
Comment (orionw) [inline submission/pipelines/generate_pipelines.py]: This doesn't come pre-installed.  Do we want a requirements.txt file?

Comment (bjschoenfeld) [reply submission/pipelines/generate_pipelines.py]: sure

File: submission/pipelines/generate_pipelines.py [hunk 5]
```diff
@@ -10,9 +11,12 @@
 from byudml.metafeature_extraction.metafeature_extraction import MetafeatureExtractor
 from byudml import __imputer_version__, __imputer_path__,  __metafeature_version__,  __metafeature_path__
 import sys
-sys.path.append('./submission')
-from utils import get_new_d3m_path, clear_directory, write_pipeline_for_submission
+sys.path.append('.')
```
Comment (bjschoenfeld) [inline submission/pipelines/generate_pipelines.py]: do we need all of `.` or is `./submission` sufficient?

Comment (orionw) [reply submission/pipelines/generate_pipelines.py]: When I ran this with './submission' it didn't run correctly so I changed it back to what I had.  I think all we need that for is to import utils.py though

Comment (bjschoenfeld) [reply submission/pipelines/generate_pipelines.py]: interesting...

**Pull request merged**